sql/core/src/main/scala/org/apache/spark/sql/execution/command/AnalyzeTableCommand.scala

```diff
@@ -88,24 +85,54 @@ case class AnalyzeTableCommand(tableName: String) extends RunnableCommand {
             }
           }.getOrElse(0L)
 
-        // Update the Hive metastore if the total size of the table is different than the size
-        // recorded in the Hive metastore.
-        // This logic is based on org.apache.hadoop.hive.ql.exec.StatsTask.aggregateStats().
-        if (newTotalSize > 0 && newTotalSize != oldTotalSize) {
-          sessionState.catalog.alterTable(
-            catalogTable.copy(
-              properties = relation.catalogTable.properties +
-                (AnalyzeTableCommand.TOTAL_SIZE_FIELD -> newTotalSize.toString)))
-        }
+        updateTableStats(
+          catalogTable,
+          oldTotalSize = catalogTable.stats.map(_.sizeInBytes.toLong).getOrElse(0L),
+          oldRowCount = catalogTable.stats.flatMap(_.rowCount.map(_.toLong)).getOrElse(-1L),
+          newTotalSize = newTotalSize)
+
+      // data source tables have been converted into LogicalRelations
+      case logicalRel: LogicalRelation if logicalRel.catalogTable.isDefined =>
+        val table = logicalRel.catalogTable.get
+        updateTableStats(
+          table,
+          oldTotalSize = table.stats.map(_.sizeInBytes.toLong).getOrElse(0L),
+          oldRowCount = table.stats.flatMap(_.rowCount.map(_.toLong)).getOrElse(-1L),
```

Comment:
so we always calculate `oldTotalSize` and `oldRowCount` by reading the `stats` field in `CatalogTable`, I think  `updateTableStats` only need to take one parameter, the `CatalogTable`, and we can calculate `oldTotalSize` and `oldRowCount` in the method.